mm2src/mm2_event_stream/src/streamer.rs

```diff
@@ -0,0 +1,232 @@
+
+        // If the handler takes no input data, return `None` for the data sender.
+        if any::TypeId::of::<Self::DataInType>() == any::TypeId::of::<NoDataIn>() {
+            Ok((tx_shutdown, None))
+        } else {
+            Ok((tx_shutdown, Some(any_data_sender)))
+        }
+    }
+}
+
+/// A wrapper around `StreamingManager` to only expose the `broadcast` method.
+pub struct Broadcaster(StreamingManager);
+
+impl Broadcaster {
+    pub fn new(inner: StreamingManager) -> Self { Self(inner) }
+
+    pub fn broadcast(&self, event: Event) { self.0.broadcast(event); }
+}
+
+#[cfg(test)]
+pub mod test_utils {
+    use super::*;
+
+    use common::executor::Timer;
+    use serde_json::json;
+
+    /// A test event streamer that broadcasts an event periodically.
+    /// Broadcasts `json!("hello")` every tenth of a second.
+    pub struct PeriodicStreamer;
+
+    #[async_trait]
+    impl EventStreamer for PeriodicStreamer {
+        type DataInType = NoDataIn;
+
+        fn streamer_id(&self) -> String { "periodic_streamer".to_string() }
+
+        async fn handle(
+            self,
+            broadcaster: Broadcaster,
+            ready_tx: oneshot::Sender<Result<(), String>>,
+            _: impl StreamHandlerInput<Self::DataInType>,
+        ) {
+            ready_tx.send(Ok(())).unwrap();
+            loop {
+                broadcaster.broadcast(Event::new(self.streamer_id(), json!("hello")));
+                Timer::sleep(0.1).await;
+            }
+        }
+    }
+
+    /// A test event streamer that broadcasts an event whenever it receives a new message through `data_rx`.
+    pub struct ReactiveStreamer;
+
+    #[async_trait]
+    impl EventStreamer for ReactiveStreamer {
+        type DataInType = String;
+
+        fn streamer_id(&self) -> String { "reactive_streamer".to_string() }
+
+        async fn handle(
+            self,
+            broadcaster: Broadcaster,
+            ready_tx: oneshot::Sender<Result<(), String>>,
+            mut data_rx: impl StreamHandlerInput<Self::DataInType>,
+        ) {
+            ready_tx.send(Ok(())).unwrap();
+            while let Some(msg) = data_rx.next().await {
+                // Just echo back whatever we receive.
+                broadcaster.broadcast(Event::new(self.streamer_id(), json!(msg)));
+            }
+        }
+    }
+
+    /// A test event streamer that fails upon initialization.
+    pub struct InitErrorStreamer;
+
+    #[async_trait]
+    impl EventStreamer for InitErrorStreamer {
+        type DataInType = NoDataIn;
+
+        fn streamer_id(&self) -> String { "init_error_streamer".to_string() }
+
+        async fn handle(
+            self,
+            _: Broadcaster,
+            ready_tx: oneshot::Sender<Result<(), String>>,
+            _: impl StreamHandlerInput<Self::DataInType>,
+        ) {
+            // Fail the initialization and stop.
+            ready_tx.send(Err("error".to_string())).unwrap();
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
It's worth to extend this coverage to WASM target (by using `cross_test` macro).